# 1.1 Introduction:
Sentimental analysis is a procedure of determining whether a piece of writing (microblogging in twitter here) is  positive, negative or neutral using several computations. It is mainly used for determining the attitude of a speaker and is also known as opinion mining.


# 1.2 Practical uses of Sentimental analysis: 
1. Inform and make operational improvements or capital expenditures. 
2. Evaluate guest likes and dislikes for your property AND your competitors’ properties.
3. Apply Sentiment Analysis for surveys. 
4. Understand your guests better ;-)


# 1.3 Objective:
To perform sentimental analysis on the tweets related to Bitcoin


# 1.4 Requirements: 
1. Python
2. Python libraies including collections, tweepy and csv
3. Aylien text analysis api


# 1.5 Preprocessing

####  1.5.1  Hashtags

A hashtag is a word or an un-spaced phrase prefixed with the hash symbol (#).
These are used to both naming subjects and phrases that are currently in
trending topics. For example, #iPad, #news

Regular Expression: `#(\w+)`

Replace Expression: `HASH_\1`

####  1.5.2  Handles

Every Twitter user has a unique username. Any thing directed towards that user
can be indicated be writing their username preceded by ‘@’. Thus, these are
like proper nouns. For example, @Apple

Regular Expression: `@(\w+)`

Replace Expression: `HNDL_\1`

####  1.5.3  URLs

Users often share hyperlinks in their tweets. Twitter shortens them using its
in-house URL shortening service, like https://t.co/1uoxlp8FwN - such links also
enables Twitter to alert users if the link leads out of its domain. From the
point of view of text classification, a particular URL is not important.
However, presence of a URL can be an important feature. Regular expression for
detecting a URL is fairly complex because of different types of URLs that can
be there, but because of Twitter’s shortening service, we can use a relatively
simple regular expression.

Regular Expression: `(http|https|ftp)://[a-zA-Z0-9\\./]+`

Replace Expression: `URL`

####  1.5.4  Emoticons

Use of emoticons is very prevalent throughout the web, more so on micro-
blogging sites. We identify the following emoticons and replace them with a
single word. Table 4 lists the emoticons we are currently detecting. All other
emoticons would be ignored.

<div style="text-align:center"> 
<table border="1">
<tr><td colspan="1" align="center">Emoticons </td><td colspan="6" align="center">Examples </td></tr>
<tr><td align="left"><tt>EMOT_SMILEY</tt>   </td><td align="left"><tt>:-)</tt>  </td><td align="left"><tt>:)</tt>   </td><td align="left"><tt>(:</tt>   </td><td align="left"><tt>(-:</tt>  </td><td align="left"><tt></tt>     </td><td align="left"><tt></tt> </td></tr>
<tr><td align="left"><tt>EMOT_LAUGH</tt>    </td><td align="left"><tt>:-D</tt>  </td><td align="left"><tt>:D</tt>   </td><td align="left"><tt>X-D</tt>  </td><td align="left"><tt>XD</tt>   </td><td align="left"><tt>xD</tt>   </td><td align="left"><tt></tt> </td></tr>
<tr><td align="left"><tt>EMOT_LOVE</tt>     </td><td align="left"><tt>&lt;3</tt>    </td><td align="left"><tt>:*</tt>   </td><td align="left"><tt></tt>     </td><td align="left"><tt></tt>     </td><td align="left"><tt></tt>     </td><td align="left"><tt></tt> </td></tr>
<tr><td align="left"><tt>EMOT_WINK</tt>     </td><td align="left"><tt>;-)</tt>  </td><td align="left"><tt>;)</tt>   </td><td align="left"><tt>;-D</tt>  </td><td align="left"><tt>;D</tt>   </td><td align="left"><tt>(;</tt>   </td><td align="left"><tt>(-;</tt> </td></tr>
<tr><td align="left"><tt>EMOT_FROWN</tt>    </td><td align="left"><tt>:-(</tt>  </td><td align="left"><tt>:(</tt>   </td><td align="left"><tt>(:</tt>   </td><td align="left"><tt>(-:</tt>  </td><td align="left"><tt></tt>     </td><td align="left"><tt></tt> </td></tr>
<tr><td align="left"><tt>EMOT_CRY</tt>  </td><td align="left"><tt>:,(</tt>  </td><td align="left"><tt>:'(</tt>  </td><td align="left"><tt>:"(</tt>  </td><td align="left"><tt>:((</tt>  </td><td align="left"><tt></tt>     </td><td align="left"><tt></tt> </td></tr></table>


<div style="text-align:center">Table 4: List of Emoticons</div>
<a id="tab:emot">
</a>
</div>

####  1.5.5  Punctuations

Although not all Punctuations are important from the point of view of
classification but some of these, like question mark, exclamation mark can
also provide information about the sentiments of the text. We replace every
word boundary by a list of relevant punctuations present at that point. Table
5 lists the punctuations currently identified. We also remove any single
quotes that might exist in the text.

<div style="text-align:center"> 
<table border="1">
<tr><td colspan="1" align="center">Punctuations </td><td colspan="2" align="center">Examples </td></tr>
<tr><td align="left"><tt>PUNC_DOT</tt> </td><td align="left"><tt>.</tt> </td><td align="left"><tt></tt> </td></tr>
<tr><td align="left"><tt>PUNC_EXCL</tt> </td><td align="left"><tt>!</tt> </td><td align="left"><tt>¡</tt> </td></tr>
<tr><td align="left"><tt>PUNC_QUES</tt> </td><td align="left"><tt>?</tt> </td><td align="left"><tt>¿</tt> </td></tr>
<tr><td align="left"><tt>PUNC_ELLP</tt> </td><td align="left"><tt>...</tt> </td><td align="left"><tt>…</tt> </td></tr></table>


<div style="text-align:center">Table 5: List of Punctuations</div>
<a id="tab:punc">
</a>
</div>

####  1.6.6  Repeating Characters

People often use repeating characters while using colloquial language, like
"I’m in a hurrryyyyy", "We won, yaaayyyyy!" As our final pre-processing step,
we replace characters repeating more than twice as two characters.

Regular Expression: `(.)\1{1,}`

Replace Expression: `\1\1`

##### source for 1.5: https://github.com/ayushoriginal/Sentiment-Analysis-Twitter/blob/master/README.md

# 1.6 Classifier used:

On little research and testing it was found that <b>SVM, Naive Bayes and maximum entropy </b> are one of the best supervised machine learning algorithms for sentimental analysis. 

On furthur research, I realised that <b>SVM</b> is superior as compared to naive bayes and maximum entropy.

That's why SVM classifier is used inorder to achieve more accuracy.

# 1.6 About training data:

The training data was collected by polling the Twitter API every four hours from April to December 2012 in search of tweets with either a positive word hashtag or a negative word hash- tag.A set of 775,000 tweets were used to generate a large word–sentiment association lexicon.Out of 775,000 tweets, 8,891 tweets were randomly selected for the training of SVM classifier. I contacted the team who collected the training data and leared that it is <b>Humanly Annotated</b>.

# 1.7 Sentimental Analysis of Tweets about Bitcoin
### 1.7.1 CSV File Sample:-

In [1]:
import pandas as pd
from IPython.display import display
from IPython.display import Image
sample = pd.read_csv('final_data.csv')
neutral = 0
positive = 0
negative = 0
for i in sample['sentiment']:
    if i == 'neutral':
        neutral+=1
    elif i == 'positive':
        positive+=1
    elif i == 'negative':
        negative+=1
        
pd.options.display.max_columns = None
sample.head(10)

tweets sentiment
0  RT @OpenPlatformICO: Move over Bitpay: How OPE...   neutral
1  In the City, a thriving metropolis encompassin...   neutral
2  RT @Pecun_io: OUR ADVISOR MR. DAVID DRAKE ON F...   neutral
3  RT @witchaindev: Airdrop 🍬🍬#100 WIT tokens（val...  positive
4  In the City, a thriving metropolis encompassin...   neutral
5  RT @Bleeding_Crypto: STOP! There is no big sur...   neutral
6  @im_uname @BTCNewsUpdates @SeatacBCH @KingDevo...  negative
7  RT @MarketWatch: The era of overnight bitcoin ...   neutral
8  Coinbase to Let Users Withdraw Funds from Bitc...   neutral
9  RT @escobar_inc: 【Dietbitcoin】Fork of Bitcoin ...   neutral

### 1.7.2 Analysis on Pie Chart:-

In [2]:
import plotly
import plotly.offline as py
import plotly.graph_objs as go
py.init_notebook_mode(connected=True)

labels = ['Positive', 'Negative', 'Neutral']
values = [positive,negative,neutral]
colors = ['#30a832', '#fc2c19', '#fcd219']
print (positive+negative+neutral, positive, negative, neutral)
domains = [
    {'x': [7.0, 7.0], 'y': [7.0, 7.0]}
]
traces = []

for domain in domains:
    trace = go.Pie(labels = labels,
                   values = values,
                   domain = domain,
                   hoverinfo = 'label+percent',
                  marker=dict(colors=colors, 
                           line=dict(color='#000000', width=3)))
    traces.append(trace)

layout = go.Layout(
                   autosize = True,
                   title = 'Tweets Analysed on Bitcoin')
fig = go.Figure(data = traces, layout = layout)
py.offline.iplot(fig, show_link = False)

157805 38375 8052 111378


### 1.7.3 Description of Chart:
Chart above represents the tweets on <b>Bitcoin</b> posted between  <b>4-08-2018 23:59:59</b> and <b>5-08-2018 23:59:59</b> in the geogrephic region of <b>39.8,-95.583068847656,2500km</b> which covers all the states of America and major part of Canada. 

On summing up, there are 157805 in total out of which  <b>38375</b> are positive, <b>8052</b> are negative and <b>111378</b> are neutral.

## 1.8 Observations:
1. Two classifiers i.e Majority classifier and SVM classifier was used. Peak accuracy of Majority classifier was 38.38% while accuracy of SVM classifier was <b>88.93</b> percent.
2. The classifier only works on tweets in english language and predicts neutral on rest of the languages. Hence only english tweets were used for predictions and other tweets were ommited.
3. Sentimental analysis on tweets can also be performed by using gradient boosting models yet little more research still required to be done.
4. I can only mine <b>72,000</b> tweets per hour as per the lastest tems and conditions of twitter and was able to achieve this number using few trciks.  

## 1.9 Conclusion:
On the basis of research done until now, SVM has proved to be the best in terms of stability and peak accuracy of <b>88.93%</b>. 